File: backend/src/lib/place-checkin-geo.ts
/** Check-in geofence: distance <= max(50m, GPS accuracy), accuracy <= 50m. */
export const PLACE_CHECKIN_MAX_ACCURACY_M = 50;
export const PLACE_CHECKIN_BASE_RADIUS_M = 50;

export function checkInAllowedRadiusM(accuracyM: number): number {
  return Math.max(PLACE_CHECKIN_BASE_RADIUS_M, accuracyM);
}

export function isWithinCheckInRadius(
  distanceM: number,
  accuracyM: number,
): boolean {
  if (!Number.isFinite(accuracyM) || accuracyM <= 0) return false;
  if (accuracyM > PLACE_CHECKIN_MAX_ACCURACY_M) return false;
  return distanceM <= checkInAllowedRadiusM(accuracyM);
}

File: mobile/lib/features/places/data/utils/place_checkin_geo.dart
import 'package:geolocator/geolocator.dart';

/// Check-in geofence: distance <= max(10m, GPS accuracy), accuracy <= 25m.
abstract final class PlaceCheckInGeo {
  static const maxAccuracyM = 25.0;
  static const baseRadiusM = 10.0;

  static double allowedRadiusM(double accuracyM) {
    if (accuracyM <= 0 || accuracyM > maxAccuracyM) return 0;
    return accuracyM > baseRadiusM ? accuracyM : baseRadiusM;
  }

  static bool isWithinRadius({
    required double distanceM,
    required double accuracyM,
  }) {
    if (accuracyM <= 0 || accuracyM > maxAccuracyM) return false;
    return distanceM <= allowedRadiusM(accuracyM);
  }

  static double distanceM({
    required double userLat,
    required double userLng,
    required double placeLat,
    required double placeLng,
  }) {
    return Geolocator.distanceBetween(userLat, userLng, placeLat, placeLng);
  }
}
